### PR TITLE
Implement the arena allocator.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
A part of #2511

### CODE REVIEW

**Comment by cangfengzhs** [comment]
Why not use the arena in folly ?

**Comment by Shylock-Hg** [comment]
> Why not use the arena in folly ?

I introduce some optimization and gain better performance than folly arena in the benchmark.

File: src/common/base/Arena.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    allocatedSize_ = 0;
+#endif
+    availableSize_ = 0;
+    currentPtr_ = nullptr;
+  }
+
+  // The CPU access memory with the alignment,
+  // So construct object from alignment address will reduce the CPU access count then
+  // speed up read/write
+  void *allocateAligned(const std::size_t alloc);
```
Comment (critical27) [inline src/common/base/Arena.h:38]: Consider only provide an interface of `allocate`. It should not be caller's duty to provide an aligned size. When user call `allocate(not aligned size)`, you can round up the size and call `allocateAligned`.

Or provide both `allocateAligned` and `allocate` LGTM as well.

File: src/common/base/Arena.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  // replace the modulo operation by bit and
+  static_assert(kAlignment && !(kAlignment & (kAlignment - 1)), "Align must be power of 2.");
+  const std::size_t pad =
+      kAlignment - (reinterpret_cast<uintptr_t>(currentPtr_) & (kAlignment - 1));
+  const std::size_t consumption = alloc + pad;
+  if (consumption > kMaxChunkSize) {
+    DLOG(FATAL) << "Arena can't allocate so large memory.";
+    return nullptr;
+  }
+  if (consumption <= availableSize_) {
```
Comment (critical27) [inline src/common/base/Arena.cpp:24]: We'd better add a LIKELY here

File: src/common/base/Arena.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    allocatedSize_ += consumption;
+#endif
+    availableSize_ -= consumption;
+    return ptr;
+  } else {
+    newChunk(std::max(alloc, kMinChunkSize));
+    // The new operator will allocate the aligned memory
+    DCHECK_EQ(reinterpret_cast<uintptr_t>(currentPtr_) & (kAlignment - 1), 0);
+    void* ptr = currentPtr_;
+    currentPtr_ += alloc;
```
Comment (critical27) [inline src/common/base/Arena.cpp:37]: `currentPtr_` need to add pad as well, otherwise, upcoming allocated won't be aligned?

**Review by critical27 — COMMENTED** [review]
A very good start! 

BTW, what do we have in the `ObjPool` for now besides AST?

File: src/common/base/Arena.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply src/common/base/Arena.cpp:37]: The `newChunk` method will allocate the aligned memory so we don't need to compute alignment here.

**Comment by Shylock-Hg** [comment]
> A very good start!
> 
> BTW, what do we have in the `ObjPool` for now besides AST?

Only AST and Plan related objects.

File: src/common/base/Arena.h (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/base/Arena.h:38]: The user don't need to provide aligned size, the method only accept one parameter which is allocated size.

Comment (critical27) [reply src/common/base/Arena.h:38]: 👌 , my bad

File: src/common/base/Arena.cpp (hunk 3 above)
Comment (critical27) [reply src/common/base/Arena.cpp:37]: 👌 , I misunderstood previously.

File: src/common/base/Arena.cpp (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/base/Arena.cpp]: Added.

**Pull request closed without merging**